src/hotspot/share/opto/escape.cpp

```diff
@@ -387,6 +398,330 @@ bool ConnectionGraph::compute_escape() {
+        }
+      }
+
+      for (uint target_phi_idx = 0; target_phi_idx < target_phis.size(); ++target_phi_idx) {
+        Node* target_phi = target_phis.at(target_phi_idx);
+        PointsToNode* target_phi_ptn = ptnode_adr(target_phi->_idx);
+
+        if (reduce_this_phi(target_phi->as_Phi())) {
+          remove_phi_node(target_phi_ptn);
+          number_of_reductions++;
+        } else {
+          assert(false, "Failed to create ReducedAllocationMerge");
+          _compile->record_failure(C2Compiler::retry_no_reduce_allocation_merges());
+          return -1;
+        }
+      }
+    }
+
+    for (DUIterator_Fast imax, i = candidate_region->fast_outs(imax); i < imax; i++) {
+      Node* m = candidate_region->fast_out(i);
+      ideal_nodes.push(m);
+    }
+  }
+
+  _igvn->set_delay_transform(prev_delay_transform);
+
+  return number_of_reductions;
+}
+
+const Node* ConnectionGraph::come_from_allocate(const Node* n) const {
+  int max_iterations = 100;
+  while (--max_iterations > 0) {
+    switch (n->Opcode()) {
+      case Op_CastPP:
+      case Op_CheckCastPP:
+      case Op_EncodeP:
+      case Op_EncodePKlass:
+      case Op_DecodeN:
+      case Op_DecodeNKlass:
+        n = n->in(1);
+        break;
+      case Op_Proj:
+        n = n->in(0);
+        break;
+      case Op_Parm:
+      case Op_GetAndSetN:
+      case Op_GetAndSetP:
+      case Op_LoadP:
+      case Op_LoadN:
+      case Op_LoadKlass:
+      case Op_LoadNKlass:
+      SHENANDOAHGC_ONLY(case Op_ShenandoahLoadReferenceBarrier:)
+      SHENANDOAHGC_ONLY(case Op_ShenandoahIUBarrier:)
+      case Op_ConP:
+      case Op_ConN:
+      case Op_CreateEx:
+      case Op_AllocateArray:
+      case Op_Phi:
+        return NULL;
+      case Op_Allocate:
+        return n;
+      default:
+        if (n->is_Call()) {
+          return NULL;
+        }
+        assert(false, "Should not reach here. Unmatched %d %s", n->_idx, n->Name());
+        return NULL;
+    }
+  }
+
+  return NULL;
+}
+
+bool ConnectionGraph::is_read_only(Node* merge_phi_region, Node* base) const {
+  Unique_Node_List worklist;
+  worklist.push(base);
+
+  for (uint next = 0; next < worklist.size(); ++next) {
+    Node* n = worklist.at(next);
+
+    for (DUIterator_Fast imax, i = n->fast_outs(imax); i < imax; i++) {
+      Node* m = n->fast_out(i);
+
+      switch (m->Opcode()) {
+        case Op_CastPP:
+        case Op_CheckCastPP:
+        case Op_EncodeP:
+        case Op_EncodePKlass:
+        case Op_DecodeN:
+        case Op_DecodeNKlass:
+          worklist.push(m);
+          break;
+      }
+
+      if (m->is_AddP()) {
+        for (DUIterator_Fast imax, i = m->fast_outs(imax); i < imax; i++) {
+          Node* child = m->fast_out(i);
+
+          if (child->is_Store()) {
+            assert(child->in(0) != NULL || m->in(0) != NULL, "No control for store or AddP.");
+            if (_igvn->is_dominator(merge_phi_region, child->in(0) != NULL ? child->in(0) : m->in(0))) {
```

Comment:
`PhaseIterGVN::is_dominator()` is a conservative approximation and can result in false negatives.